### PR TITLE
RUBY-2014 Send logging messages from libmongocrypt to Mongo::Logger

### PR DESCRIPTION
**Pull request by egiurleo**
Also change the context api (data key context, explicit encryption context, explicit decryption context) to be more consistent with the rest of the `Crypt` API that's been built so far.

### CODE REVIEW

File: spec/mongo/crypt/explicit_encryption_context_spec.rb [hunk 1]
```diff
@@ -79,6 +78,40 @@
...
+          # option; none of the pre-built packages on Evergreen have been built with logging enabled.
+          #
+          # It is still useful to be able to run these tests locally to confirm that logging is working
+          # while debugging any problems.
+          #
+          # For now, skip this test by default and revisit once we have determined how we want to
+          # package libmongocrypt with the Ruby driver.
+          skip "These tests require libmongocrypt to be built with the '-DENABLE_TRACE=ON' cmake option." +
+            " They also require the MONGOCRYPT_TRACE environment variable to be set to 'ON'."
+        end
```
Comment (egiurleo) [inline spec/mongo/crypt/explicit_encryption_context_spec.rb]: @p-mongo let me know if you're okay leaving the tests like this for now.

File: lib/mongo/crypt/handle.rb [hunk 2]
```diff
@@ -52,6 +52,31 @@ def ref
...
+          when :fatal
+            logger.fatal(msg)
+          when :error
+            logger.error(msg)
+          when :warning
+            logger.warn(msg)
+          when :info
+            logger.info(msg)
+          when :trace
+            logger.debug(msg)
```
Comment (p-mongo) [inline lib/mongo/crypt/handle.rb]: Would it make sense to have the level symbols match the Ruby logger methods, to simplify the mapping?

File: lib/mongo/crypt/handle.rb [hunk 3]
```diff
@@ -52,6 +52,31 @@ def ref
 
       private
 
+      # Send the logs from libmongocrypt to the Mongo::Logger
+      def set_logger
+        @log_callback = Proc.new do |level, msg|
+          logger = Mongo::Logger.logger
```
Comment (p-mongo) [inline lib/mongo/crypt/handle.rb]: This takes the global logger, not the client's logger for any particular client. This is an existing issue in the driver affecting a number of call sites, but is the client's logger conveniently available here? Or, is the logger here intended to be the global one?

File: spec/mongo/crypt/explicit_encryption_context_spec.rb (hunk 1 above)
Comment (p-mongo) [reply spec/mongo/crypt/explicit_encryption_context_spec.rb]: I think it's fine, maybe link to the jira ticket for packaging if you are going to make other changes.

File: lib/mongo/crypt/handle.rb (hunk 3 above)
Comment (egiurleo) [reply lib/mongo/crypt/handle.rb]: Nope, you're right, I meant it to be the client's logger. I'll fix it. Thanks for pointing that out.

File: lib/mongo/crypt/handle.rb (hunk 2 above)
Comment (egiurleo) [reply lib/mongo/crypt/handle.rb]: That's awesome, thank you.

File: lib/mongo/crypt/handle.rb [hunk 4]
```diff
@@ -22,23 +22,28 @@ module Crypt
...
+      #
+      # @option [ Logger ] :logger A Logger object to which libmongocrypt logs
+      #   will be sent
       #
       # There will be more arguemnts to this method once automatic encryption is introduced.
-      def initialize(kms_providers)
+      def initialize(kms_providers, options={})
         # FFI::AutoPointer uses a custom release strategy to automatically free
         # the pointer once this object goes out of scope
+        @logger = options[:logger]
```
Comment (egiurleo) [inline lib/mongo/crypt/handle.rb]: @p-mongo I changed the API to accept a logger as an option, with the intention of passing it down from the Client or ClientEncryption object. Does that seem reasonable?

**Pull request merged**